File: apps/api/internal/auth/apikey.go
package auth

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"strings"
	"time"

	"github.com/gin-gonic/gin"

	"petrichor/api/internal/db"
	httpx "petrichor/api/internal/httpx"
)

// Agent API Key 体系，复刻 src/server/agent/api-key.ts。

var AgentApiKeyScopes = []string{
	"article:write",
	"article:delete",
	"doc:read",
	"qa:read",
	"share:write",
	"ai:write",
	"wiki:read",
	"wiki:write",
}

var scopeSet = func() map[string]struct{} {
	m := make(map[string]struct{}, len(AgentApiKeyScopes))
	for _, s := range AgentApiKeyScopes {
		m[s] = struct{}{}
	}
	return m
}()

// AgentAuthContext Agent Key 鉴权上下文。
type AgentAuthContext struct {
	ApiKeyID   int64
	KeyPrefix  string
	Scopes     []string
	UserID     int64
}

const agentKeyCtx = "petrichor.agentAuth"

// GenerateAgentApiKey 生成 ptc_live_ 前缀 key。
func GenerateAgentApiKey() string {
	buf := make([]byte, 32)
	_, _ = rand.Read(buf)
	return "ptc_live_" + base64.RawURLEncoding.EncodeToString(buf)
}

// ToAgentApiKeyPrefix 前 16 位 + "..."。
func ToAgentApiKeyPrefix(apiKey string) string {
	trimmed := strings.TrimSpace(apiKey)
	return trimmed[:16] + "..."
}

// ExtractAgentBearerToken Authorization Bearer 或 x-petrichor-api-key。
func ExtractAgentBearerToken(c *gin.Context) string {
	authz := strings.TrimSpace(c.GetHeader("Authorization"))
	if len(authz) > 7 && strings.EqualFold(authz[:7], "Bearer ") {
		token := strings.TrimSpace(authz[7:])
		if token != "" {
			return token
		}
	}
	return strings.TrimSpace(c.GetHeader("x-petrichor-api-key"))
}

// ParseAgentScopes 解析 scopes JSON 数组。
func ParseAgentScopes(raw *string) []string {
	if raw == nil || strings.TrimSpace(*raw) == "" {
		return []string{}
	}
	var arr []any
	if err := json.Unmarshal([]byte(*raw), &arr); err != nil {
		return []string{}
	}
	scopes := make([]string, 0, len(arr))
	for _, item := range arr {
		s, ok := item.(string)
		if ok {
			if _, valid := scopeSet[s]; valid {
				scopes = append(scopes, s)
			}
		}
	}
	return scopes
}

// HasAgentScope 校验权限。
func HasAgentScope(scopes []string, required string) bool {
	for _, s := range scopes {
		if s == required {
			return true
		}
	}
	return false
}

func storeAgentCtx(c *gin.Context, actx *AgentAuthContext) { c.Set(agentKeyCtx, actx) }

// AgentAuth 从上下文取 Agent 鉴权信息。
func AgentAuth(c *gin.Context) *AgentAuthContext {
	v, _ := c.Get(agentKeyCtx)
	actx, _ := v.(*AgentAuthContext)
	return actx
}

// RequireAgentKey Agent API Key 鉴权中间件。
func RequireAgentKey(requiredScope string) gin.HandlerFunc {
	return func(c *gin.Context) {
		apiKey := ExtractAgentBearerToken(c)
		if apiKey == "" {
			httpx.ErrorJSON(c, 401, "缺少 Agent API Key")
			return
		}
		keyHash := HashAgentApiKey(apiKey)
		ctx := context.Background()
		pool := db.Pool()
		var (
			id        int64
			userID    int64
			keyPrefix string
			scopesRaw *string
			expiresAt *time.Time
		)
		err := pool.QueryRow(ctx,
			`SELECT id, user_id, key_prefix, scopes_json, expires_at
			 FROM petrichor_agent_api_key
			 WHERE key_hash = $1 AND revoked_at IS NULL AND (expires_at IS NULL OR expires_at > now())
			 LIMIT 1`, keyHash).
			Scan(&id, &userID, &keyPrefix, &scopesRaw, &expiresAt)
		if err != nil {
			httpx.ErrorJSON(c, 401, "Agent API Key 无效或已过期")
			return
		}
		_, _ = pool.Exec(ctx,
			`UPDATE petrichor_agent_api_key SET last_used_at = now(), updated_at = now() WHERE id = $1`, id)

		scopes := ParseAgentScopes(scopesRaw)
		if requiredScope != "" && !HasAgentScope(scopes, requiredScope) {
			httpx.ErrorJSON(c, 403, "当前 API Key 缺少 "+requiredScope+" 权限")
			return
		}
		storeAgentCtx(c, &AgentAuthContext{
			ApiKeyID:  id,
			KeyPrefix: keyPrefix,
			Scopes:    scopes,
			UserID:    userID,
		})
		c.Next()
	}
}

File: apps/web/src/server/agent/api-key.ts
import { createHash, randomBytes } from "node:crypto"
import { and, eq, gt, isNull, or } from "drizzle-orm"
import type { AppRequest } from "@/server/http/request"
import { z } from "zod"
import { getDb } from "@/server/db/client"
import { agentApiKeys, type AgentApiKeyRecord } from "@/server/db/schema"
import { badRequest, forbidden, unauthorized } from "@/server/http/response"

export const AGENT_API_KEY_SCOPES = [
    "article:write",
    "article:delete",
    "doc:read",
    "qa:read",
    "share:write",
    "ai:write",
    "wiki:read",
    "wiki:write",
    "external:read",
] as const

export type AgentApiKeyScope = typeof AGENT_API_KEY_SCOPES[number]

export type AgentAuthContext = {
    apiKey: AgentApiKeyRecord
    scopes: AgentApiKeyScope[]
    userId: number
}

const scopeSet = new Set<string>(AGENT_API_KEY_SCOPES)
export const DEFAULT_AGENT_API_KEY_SCOPES = AGENT_API_KEY_SCOPES.filter(
    (scope): scope is Exclude<AgentApiKeyScope, "external:read"> => scope !== "external:read",
)

export const agentApiKeyCreateSchema = z.object({
    expiresAt: z.string().datetime().optional().nullable(),
    name: z.string().trim().min(1).max(80).default("Agent Skill Key"),
    scopes: z.array(z.enum(AGENT_API_KEY_SCOPES)).optional().default([...DEFAULT_AGENT_API_KEY_SCOPES]),
})

export const agentApiKeyUpdateSchema = z.object({
    id: z.union([z.string(), z.number()]).transform((value, ctx) => {
        const id = Number(value)
        if (!Number.isInteger(id) || id <= 0) {
            ctx.addIssue({ code: "custom", message: "ID 必须是正整数" })
            return z.NEVER
        }
        return id
    }),
    scopes: z.array(z.enum(AGENT_API_KEY_SCOPES)).min(1),
})

export const agentApiKeyRevokeSchema = z.object({
    id: z.union([z.string(), z.number()]).transform((value, ctx) => {
        const raw = String(value).trim()
        if (!/^\d+$/.test(raw) || Number(raw) <= 0) {
            ctx.addIssue({ code: "custom", message: "ID 必须是正整数" })
            return z.NEVER
        }
        return Number(raw)
    }),
})

export function generateAgentApiKey() {
    return `ptc_live_${randomBytes(32).toString("base64url")}`
}

export function hashAgentApiKey(apiKey: string) {
    return createHash("sha256").update(apiKey, "utf8").digest("hex")
}

export function toAgentApiKeyPrefix(apiKey: string) {
    const trimmed = apiKey.trim()
    return `${trimmed.slice(0, 16)}...`
}

export function extractAgentBearerToken(request: Request) {
    const authorization = request.headers.get("authorization")?.trim() ?? ""
    const bearerMatch = authorization.match(/^Bearer\s+(.+)$/i)
    if (bearerMatch?.[1]?.trim()) {
        return bearerMatch[1].trim()
    }

    const apiKeyHeader = request.headers.get("x-petrichor-api-key")?.trim()
    return apiKeyHeader || null
}

export function parseAgentScopes(raw: string | null | undefined): AgentApiKeyScope[] {
    if (!raw?.trim()) return []
    try {
        const parsed = JSON.parse(raw) as unknown
        if (!Array.isArray(parsed)) return []
        return parsed.filter((item): item is AgentApiKeyScope => typeof item === "string" && scopeSet.has(item))
    } catch {
        return []
    }
}

export function hasAgentScope(scopes: AgentApiKeyScope[], required: AgentApiKeyScope) {
    return scopes.includes(required)
}

export function requireAgentScope(context: AgentAuthContext, required: AgentApiKeyScope) {
    if (!hasAgentScope(context.scopes, required)) {
        throw forbidden(`当前 API Key 缺少 ${required} 权限`)
    }
}

export async function authenticateAgentRequest(request: AppRequest): Promise<AgentAuthContext> {
    const apiKey = extractAgentBearerToken(request)
    if (!apiKey) {
        throw unauthorized("缺少 Agent API Key")
    }

    const keyHash = hashAgentApiKey(apiKey)
    const now = new Date()
    const db = getDb()
    const [record] = await db
        .select()
        .from(agentApiKeys)
        .where(and(
            eq(agentApiKeys.keyHash, keyHash),
            isNull(agentApiKeys.revokedAt),
            or(isNull(agentApiKeys.expiresAt), gt(agentApiKeys.expiresAt, now)),
        ))
        .limit(1)

    if (!record) {
        throw unauthorized("Agent API Key 无效或已过期")
    }

    await db
        .update(agentApiKeys)
        .set({ lastUsedAt: now, updatedAt: now })
        .where(eq(agentApiKeys.id, record.id))

    return {
        apiKey: record,
        scopes: parseAgentScopes(record.scopesJson),
        userId: record.userId,
    }
}

export function parseAgentApiKeyExpiresAt(value: string | null | undefined) {
    if (!value) return null
    const date = new Date(value)
    if (Number.isNaN(date.getTime())) {
        throw badRequest("expiresAt 必须是合法时间")
    }
    if (date.getTime() <= Date.now()) {
        throw badRequest("expiresAt 必须晚于当前时间")
    }
    return date
}

export function toAgentApiKeyResponse(record: AgentApiKeyRecord) {
    return {
        id: String(record.id),
        name: record.name,
        keyPrefix: record.keyPrefix,
        scopes: parseAgentScopes(record.scopesJson),
        expiresAt: formatDate(record.expiresAt),
        lastUsedAt: formatDate(record.lastUsedAt),
        revokedAt: formatDate(record.revokedAt),
        createdAt: formatDate(record.createdAt),
        updatedAt: formatDate(record.updatedAt),
    }
}

function formatDate(value: Date | string | null | undefined): string | null {
    if (!value) return null
    return value instanceof Date ? value.toISOString() : new Date(value).toISOString()
}
